File: aws-s3/scripts-ccao-data-warehouse-us-east-1/spatial-access.R
library(arrow)
library(aws.s3)
library(dplyr)
library(here)
library(osmdata)
library(purrr)
library(sf)
library(sfarrow)
library(stringr)
library(tidyr)
source("utils.R")

# This script cleans shapefiles that represent desirable amenities, such as
# parks and hospitals
AWS_S3_RAW_BUCKET <- Sys.getenv("AWS_S3_RAW_BUCKET")
AWS_S3_WAREHOUSE_BUCKET <- Sys.getenv("AWS_S3_WAREHOUSE_BUCKET")
input_bucket <- file.path(AWS_S3_RAW_BUCKET, "spatial", "access")
output_bucket <- file.path(AWS_S3_WAREHOUSE_BUCKET, "spatial", "access")
current_year <- strftime(Sys.Date(), "%Y")

##### BIKE TRAIL #####
remote_file_bike_raw <- file.path(
  input_bucket, "bike_trail", "2021.geojson"
)
remote_file_bike_warehouse <- file.path(
  output_bucket, "bike_trail", "2021.parquet"
)

if (!aws.s3::object_exists(remote_file_bike_warehouse)) {
  tmp_file_bike <- tempfile(fileext = ".geojson")
  aws.s3::save_object(remote_file_bike_raw, file = tmp_file_bike)

  st_read(tmp_file_bike) %>%
    st_transform(4326) %>%
    rename_with(tolower) %>%
    mutate(
      geometry_3435 = st_transform(geometry, 3435)
    ) %>%
    select(-c(created_us:shape_stle)) %>%
    rename(
      speed_limit = spdlimit, on_street = onstreet, edit_date = edtdate,
      trail_width = trailwdth, trail_type = trailtype,
      trail_surface = trailsurfa
    ) %>%
    sfarrow::st_write_parquet(remote_file_bike_warehouse)
}


##### CEMETERY #####
remote_file_ceme_raw <- file.path(
  input_bucket, "cemetery", "2021.geojson"
)
remote_file_ceme_warehouse <- file.path(
  output_bucket, "cemetery", "2021.parquet"
)

if (!aws.s3::object_exists(remote_file_ceme_warehouse)) {
  tmp_file_ceme <- tempfile(fileext = ".geojson")
  aws.s3::save_object(remote_file_ceme_raw, file = tmp_file_ceme)

  st_read(tmp_file_ceme) %>%
    st_transform(4326) %>%
    rename_with(tolower) %>%
    mutate(
      geometry_3435 = st_transform(geometry, 3435)
    ) %>%
    select(
      name = cfname, address, gniscode, source, community, comment, mergeid,
      geometry, geometry_3435
    ) %>%
    sfarrow::st_write_parquet(remote_file_ceme_warehouse)
}


##### HOSPITAL #####
remote_file_hosp_raw <- file.path(
  input_bucket, "hospital", "2021.geojson"
)
remote_file_hosp_warehouse <- file.path(
  output_bucket, "hospital", "2021.parquet"
)

if (!aws.s3::object_exists(remote_file_hosp_warehouse)) {
  tmp_file_hosp <- tempfile(fileext = ".geojson")
  aws.s3::save_object(remote_file_hosp_raw, file = tmp_file_hosp)

  st_read(tmp_file_hosp) %>%
    st_transform(4326) %>%
    rename_with(tolower) %>%
    mutate(
      geometry_3435 = st_transform(geometry, 3435)
    ) %>%
    select(
      name = cfname, address, gniscode, source, community, comment, mergeid,
      geometry, geometry_3435
    ) %>%
    sfarrow::st_write_parquet(remote_file_hosp_warehouse)
}


##### PARK #####
# Switched to using OSM parks because the county-provided parks file is
# very incomplete
remote_file_park_warehouse <- file.path(
  output_bucket, "park", paste0(current_year, ".parquet")
)

if (!aws.s3::object_exists(remote_file_park_warehouse)) {

  parks <- opq("Cook County United States") %>%
    add_osm_feature(key = "leisure", value = "park") %>%
    osmdata_sf()

  cook_boundary <- st_read(
    paste0(
      "https://opendata.arcgis.com/datasets/",
      "ea127f9e96b74677892722069c984198_1.geojson"
    )
  ) %>%
    st_transform(4326)

  parks_df <- bind_rows(parks$osm_polygons, parks$osm_multipolygons) %>%
    st_make_valid() %>%
    st_cast("MULTIPOLYGON") %>%
    st_transform(4326) %>%
    filter(st_is_valid(.)) %>%
    select(osm_id, name, geometry) %>%
    mutate(geometry_3435 = st_transform(geometry, 3435)) %>%
    filter(
      as.logical(st_intersects(
        geometry,
        cook_boundary
      ))
    )

  st_write_parquet(parks_df, remote_file_park_warehouse, compression = "snappy")
}


##### INDUSTRIAL CORRIDOR #####
remote_file_indc_raw <- file.path(
  input_bucket, "industrial_corridor", "2013.geojson"
)
remote_file_indc_warehouse <- file.path(
  output_bucket, "industrial_corridor", "2013.parquet"
)

if (!aws.s3::object_exists(remote_file_indc_warehouse)) {
  tmp_file_indc <- tempfile(fileext = ".geojson")
  aws.s3::save_object(remote_file_indc_raw, file = tmp_file_indc)

  st_read(tmp_file_indc) %>%
    st_transform(4326) %>%
    rename_with(tolower) %>%
    mutate(
      geometry_3435 = st_transform(geometry, 3435)
    ) %>%
    select(
      name, region,
      num = no, hud_qualif, acres,
      geometry, geometry_3435
    ) %>%
    sfarrow::st_write_parquet(remote_file_indc_warehouse)
}

##### WALKABILITY #####
# data dictionary located on page 7 at
# https://datahub.cmap.illinois.gov/dataset/aac0d840-77b4-4e88-8a26-7220ac6c588f/resource/7f0d890f-e678-46f8-9a6e-8d0b6ad04ae7/download/WalkabilityMethodology.pdf
remote_file_walk_raw <- file.path(
  input_bucket, "walkability", "2017.geojson"
)
remote_file_walk_warehouse <- file.path(
  output_bucket, "walkability", "2017.parquet"
)

if (!aws.s3::object_exists(remote_file_walk_warehouse)) {
  tmp_file_walk <- tempfile(fileext = ".geojson")
  aws.s3::save_object(remote_file_walk_raw, file = tmp_file_walk)

  temp <- st_read(tmp_file_walk) %>%
    st_transform(4326) %>%
    rename_with(tolower) %>%
    rename_with(~ gsub("sc$|sco|scor|score", "_score", .x)) %>%
    rename_with(~ "walk_num", contains("subzone")) %>%
    rename(walkability_rating = walkabilit, amenities_score = amenities, transitaccess = transitacc) %>%
    standardize_expand_geo() %>%
    select(-contains("shape")) %>%
    sfarrow::st_write_parquet(remote_file_walk_warehouse)
}

File: aws-s3/scripts-ccao-data-warehouse-us-east-1/spatial-access.py
import datetime
import os
from datetime import datetime
import tempfile
import boto3
import geopandas as gpd
import pyarrow.parquet as pq

# Set up AWS credentials
AWS_S3_RAW_BUCKET = os.getenv("AWS_S3_RAW_BUCKET")
AWS_S3_WAREHOUSE_BUCKET = os.getenv("AWS_S3_WAREHOUSE_BUCKET")
s3 = boto3.client("s3")

current_year = datetime.now().strftime("%Y")


# Helper function to save to S3
def upload_to_s3(local_file, bucket, s3_key):
    s3.upload_file(local_file, bucket, s3_key)


# Helper function to check if an S3 object exists
def object_exists(bucket, key):
    try:
        s3.head_object(Bucket=bucket, Key=key)
        return True
    except:
        return False

# Bike Trail Data Processing
def process_bike_trail(spark_session):
    bike_key_raw = f"{AWS_S3_RAW_BUCKET}/spatial/access/bike_trail/2021.geojson"
    bike_key_warehouse = f"{AWS_S3_WAREHOUSE_BUCKET}/spatial/access/bike_trail/year=2021/part-0.parquet"

    if not object_exists(AWS_S3_WAREHOUSE_BUCKET, bike_key_warehouse):
        with tempfile.NamedTemporaryFile(suffix=".geojson") as temp:
            s3.download_file(AWS_S3_RAW_BUCKET, "spatial/access/bike_trail/2021.geojson", temp.name)
            df_bike = gpd.read_file(temp.name).to_crs(epsg=4326)
            df_bike.columns = map(str.lower, df_bike.columns)
            df_bike["geometry_3435"] = df_bike["geometry"].to_crs(epsg=3435)
            df_bike = df_bike.rename(
                columns={
                    "spdlimit": "speed_limit",
                    "onstreet": "on_street",
                    "edtdate": "edit_date",
                    "trailwdth": "trail_width",
                    "trailtype": "trail_type",
                    "trailsurfa": "trail_surface",
                }
            ).drop(columns=["created_us", "shape_stle"])

            spark_df = spark_session.createDataFrame(df_bike)
            spark_df.write.parquet(bike_key_warehouse)

# Park Data Processing
def process_parks(spark_session):
    park_key_warehouse = f"{AWS_S3_WAREHOUSE_BUCKET}/spatial/access/park/year=2021/part-0.parquet"

    if not object_exists(AWS_S3_WAREHOUSE_BUCKET, park_key_warehouse):
        with tempfile.NamedTemporaryFile(suffix=".geojson") as temp:
            cook_boundary_key = "spatial/ccao/county/2019.parquet"
            s3.download_file(AWS_S3_WAREHOUSE_BUCKET, cook_boundary_key, temp.name)
            cook_boundary = gpd.read_parquet(temp.name).to_crs(epsg=4326)

            parks_df = gpd.read_file("path/to/local/park_data.geojson").to_crs(epsg=4326)
            parks_df["geometry_3435"] = parks_df["geometry"].to_crs(epsg=3435)
            parks_df_filtered = parks_df.loc[parks_df.intersects(cook_boundary.unary_union)]

            spark_df = spark_session.createDataFrame(parks_df_filtered)
            spark_df.write.parquet(park_key_warehouse)

# Industrial Corridor Data Processing
def process_industrial_corridor(spark_session):
    indc_key_raw = f"{AWS_S3_RAW_BUCKET}/spatial/access/industrial_corridor/2013.geojson"
    indc_key_warehouse = f"{AWS_S3_WAREHOUSE_BUCKET}/spatial/access/industrial_corridor/year=2013/part-0.parquet"

    if not object_exists(AWS_S3_WAREHOUSE_BUCKET, indc_key_warehouse):
        with tempfile.NamedTemporaryFile(suffix=".geojson") as temp:
            s3.download_file(AWS_S3_RAW_BUCKET, "spatial/access/industrial_corridor/2013.geojson", temp.name)
            df_indc = gpd.read_file(temp.name).to_crs(epsg=4326)
            df_indc.columns = map(str.lower, df_indc.columns)
            df_indc["geometry_3435"] = df_indc["geometry"].to_crs(epsg=3435)
            df_indc = df_indc.rename(
                columns={
                    "name": "name",
                    "region": "region",
                    "no": "num",
                    "hud_qualif": "hud_qualif",
                }
            ).loc[:, ["name", "region", "num", "hud_qualif", "acres", "geometry", "geometry_3435"]]

            spark_df = spark_session.createDataFrame(df_indc)
            spark_df.write.parquet(indc_key_warehouse)

# Cemetery Data Processing
def process_cemetery(spark_session):
    ceme_key_raw = f"{AWS_S3_RAW_BUCKET}/spatial/access/cemetery/2021.geojson"
    ceme_key_warehouse = f"{AWS_S3_WAREHOUSE_BUCKET}/spatial/access/cemetery/year=2021/part-0.parquet"

    if not object_exists(AWS_S3_WAREHOUSE_BUCKET, ceme_key_warehouse):
        with tempfile.NamedTemporaryFile(suffix=".geojson") as temp:
            s3.download_file(AWS_S3_RAW_BUCKET, "spatial/access/cemetery/2021.geojson", temp.name)
            df_ceme = gpd.read_file(temp.name).to_crs(epsg=4326)
            df_ceme.columns = map(str.lower, df_ceme.columns)
            df_ceme["geometry_3435"] = df_ceme["geometry"].to_crs(epsg=3435)
            df_ceme = df_ceme.rename(columns={"cfname": "name"}).loc[
                :, ["name", "address", "gniscode", "source", "community", "comment", "mergeid", "geometry", "geometry_3435"]
            ]

            spark_df = spark_session.createDataFrame(df_ceme)
            spark_df.write.parquet(ceme_key_warehouse)

# Walkability Data Processing
def process_walkability(spark_session):
    walk_key_raw = f"{AWS_S3_RAW_BUCKET}/spatial/access/walkability/2017.geojson"
    walk_key_warehouse = f"{AWS_S3_WAREHOUSE_BUCKET}/spatial/access/walkability/year=2017/part-0.parquet"

    if not object_exists(AWS_S3_WAREHOUSE_BUCKET, walk_key_warehouse):
        with tempfile.NamedTemporaryFile(suffix=".geojson") as temp:
            s3.download_file(AWS_S3_RAW_BUCKET, "spatial/access/walkability/2017.geojson", temp.name)
            df_walk = gpd.read_file(temp.name).to_crs(epsg=4326)
            df_walk.columns = map(str.lower, df_walk.columns)
            df_walk.columns = [col.replace("sc", "_score") for col in df_walk.columns]
            df_walk.rename(
                columns={
                    "walkabilit": "walkability_rating",
                    "amenities": "amenities_score",
                    "transitacc": "transitaccess",
                },
                inplace=True,
            )
            df_walk = gpd.standardize_expand_geo(df_walk)
            df_walk["year"] = "2017"

            spark_df = spark_session.createDataFrame(df_walk)
            spark_df.write.parquet(walk_key_warehouse)
